Disable automatic write protection on replicas (#42450)
rails/rails: activerecord/CHANGELOG.md

> *   Disable automatic write protection on replicas

Reviewer: ```suggestion
*   Disable automatic write protection on replicas.
```

>     Write protection no longer automatically enable for replicas.

Reviewer: ```suggestion
    Write protection is no longer automatically enabled for replicas. Write protection should be enabled by the database user settings.
```

>     Now each application controls when to enable automatic write protection.

Reviewer: ```suggestion
```

*   Disable automatic write protection on replicas

    Write protection no longer automatically enable for replicas.

>     Now each application controls when to enable automatic write protection.

Reviewer 2: Do you think it's worth mentioning this on https://edgeguides.rubyonrails.org/active_record_multiple_databases.html too?

Author: I am reading now, if it was mentioned before we should probably remove it, but I don't know how/if we should mention rails not doing something you maybe didn't know it did at all before.

Reviewer 2: I had a quick read through and couldn't find any mention of this behaviour. I think `AR.while_preventing_writes` could be mentioned but it's not really in the scope of this PR.

Author: I can't find a spot that seems like it needs updating, but it is totally possible I am missing it. Let me know if I missed anything @ghiculescu 🙂

Reviewer: I would add onto the guides and explain that if you want replicas to act like replicas they need to be configured with readonly users

Reviewer 2: ^ that is alluded to, kind of

![image](https://user-images.githubusercontent.com/509837/121720136-a8c91900-caa8-11eb-8ef7-91f6a470cf6a.png)

@HParker I'll make a separate PR, because I don't think I'm making much sense here.

Author: Under the setup section we have:

> Second, the username for the writers and replicas should be different, and the replica user's permissions should be set to only read and not write.
> 
> When using a replica database, you need to add a replica: true entry to the replica in the database.yml. This is because Rails otherwise has no way of knowing which one is a replica and which one is the writer.

https://edgeguides.rubyonrails.org/active_record_multiple_databases.html#setting-up-your-application

In the guide already. Should I expand on that?

*   Disable automatic write protection on replicas.

>     Write protection is no longer automatically enabled for replicas. Write protection should be enabled by the database user settings.

Reviewer 3: Can we expand with an example what "Write protection should be enabled by the database user settings." means?

Author: Good idea @rafaelfranca, I put together https://github.com/rails/rails/pull/42529

I am not sure how much detail to add since it is not only database specific, but also configuration specific. Depending on how your primary and replica are setup you might want a totally different configuration. Ideally you would setup a local database user to be very similar to how your production users are setup.
